Merge continuous aggregate invalidations

erimatnor opened the pull request:
This change implements deduplication and merging of invalidation
entries for continuous aggregates in order to reduce the number of
reduntant entries in the continuous aggregate invalidation
log. Merging is done both when copying over entries from the
hypertable to the continuous aggregate invalidation log and when
cutting already existing invalidations in the latter log. Doing this
merging in both steps helps reduce the number of invalidations also
for the continuous aggregates that don't get refreshed by the active
refresh command.

Merging works by scanning invalidations in order of the lowest
modified value, and given this ordering it is possible to merge the
current and next entry into one large entry if they are
overlapping. This can continue until the current and next invalidation
are disjoint or there are no more invalidations to process.

Note, however, that only the continuous aggregate that gets refreshed
will be fully deduplicated. Some redundant entries might exists for
other aggregates since their entries in the continuous aggregate log
aren't cut against the refresh window.

Full deduplication for the refreshed continuous aggregate is only
possible if the continuous aggregate invalidation log is processed
last, since that also includes "old" entries. Therefore, this change
also changes the ordering of how the logs are processed. This also
makes it possible to process the hypertable invalidation log in the
first transaction of the refresh.

Closes #2179

k-rus commented:
@erimatnor Can you update the commit message and already at the beginning mention that merging happens during the copy of invalidations into cagg invalidation table?

k-rus commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -423,49 +585,54 @@ clear_cagg_invalidations_for_refresh(const CaggInvalidationState *state,
-		switch (result)
+		invalidation_entry_set_from_cagg_invalidation(&logentry, ti);
+
+		if (!IS_VALID_INVALIDATION(&mergedentry))
+			mergedentry = logentry;
```
This case is the first log entry and the entry should be removed or cut if inside the refresh window. My understanding is that it will happen on a future iteration, when try to merge fails or after the loop. Do I understand the logic correctly?

mkindahl commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -250,19 +250,178 @@ cut_invalidation_along_refresh_window(const CaggInvalidationState *state, int64
+		HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);                    \
+		type form;                                                                                 \
+		form = (type) GETSTRUCT(tuple);                                                            \
+		(entry)->hyper_id = form->hypertable_id;                                                   \
```
Well, duplicating the function does not add any value, so then it's better to keep it.

k-rus commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -250,19 +250,178 @@ cut_invalidation_along_refresh_window(const CaggInvalidationState *state, int64
+	/* Quick exit if no overlap. We need to account for entries of size 1
+	 * (lowest = greatest). */
+	if (entry->lowest_modified_value < newentry->lowest_modified_value &&
+		entry->greatest_modified_value <= newentry->lowest_modified_value)
```
It will not be sufficient when `entry->lowest_modified_value == entry->greatest_modified_value`. The only second condition will give `true`, while if both, it will be `false`.

k-rus replied:
cc @mkindahl @erimatnor or do I miss something?

k-rus replied:
And I guess there is no test for this case, when lowest and greatest of entry are equivalent, right? If so, it is good to have a test.

erimatnor replied:
I actually changed this to just

```
if (entry->greatest_modified_value < newentry->lowest_modified_value)
```

this means that an entry like [10,10] will overlap with another entry [10,10], like it should. I believe this is covered in tests.

One thing that makes this a bit confusing is that invalidations are, AFAIK, inclusive in both lowest and greatest, while, e.g., the refresh window is exclusive at the end (makes me think that maybe the refresh window should be inclusive in both ends too?).

mkindahl replied:
IMHO, having an upper end that is non-inclusive is a lot easier to work with. C++ has adopted this convention, as has a bunch of other languages for this reason.

k-rus replied:
Inclusive on both ends seems to be better for implementation and works well for invalidation semantics. E.g., [10,10) is incorrect, and it should be [10, 10.1) or something like this depending on the type.

I don't remember that it was discussed how refresh window should be treated. I find no information in the docs if the current upper limit, which is `refresh-lag`, is inclusive or exclusive. Furthermore, the refresh window is transformed into bucket widths.

mkindahl commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -250,19 +250,178 @@ cut_invalidation_along_refresh_window(const CaggInvalidationState *state, int64
+	MemSet(entry, 0, sizeof(Invalidation));
+}
+
+#define invalidation_entry_set(entry, ti, hypertable_id, type)                                     \
```
Then please make sure that it is not confused with a function call and that the parameters are not confused with fields.

erimatnor commented on tsl/src/continuous_aggs/invalidation.c:
```diff
@@ -250,19 +250,178 @@ cut_invalidation_along_refresh_window(const CaggInvalidationState *state, int64
+	do                                                                                             \
+	{                                                                                              \
+		bool should_free;                                                                          \
+		HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);                    \
```
I didn't change this because I wasn't sure why you want only some macro parameters to be all caps. Also, I haven't seem many other places where macro parameters are all caps.

mkindahl replied:
I actually prefer to have all parameters as caps, but I only changed those that had matching changes.

Granted, this convention does not seem to be used in the PG code, so this is a convention that exists in other guidelines to distinguish parameter names. If you decide to not change the parameters, I'm fine since this is not something that we have as a guideline.

Please add the parentheses though.

erimatnor commented:
@mkindahl @k-rus Thanks for the feedback. I believe I have addressed most, if not all, your comments. A heads up: I also changed how the invalidation logs are processed across the two transactions run by refresh. Since the merging of invalidations changed the order of processing logs to first hyper invalidation log, then cagg invalidation log, it is not possible to process the hyper log in the first transaction. This makes these entries visible sooner and will release the heavy lock on that table quicker. It also matches better the original materialization code.
